File: Provenance/Controller/PVStellaControllerViewController.m
//
//  PVStellaControllerViewController.m
//  Provenance
//
//  Created by James Addyman on 05/09/2013.
//  Copyright (c) 2013 James Addyman. All rights reserved.
//

#import "PVStellaControllerViewController.h"
#import <PVStella/PVStellaGameCore.h>

@interface PVStellaControllerViewController ()

@end

@implementation PVStellaControllerViewController

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    for (JSButton *button in [self.buttonGroup subviews])
    {
        if (![button isMemberOfClass:[JSButton class]])
        {
            continue; // skip over the PVButtonGroupOverlayView
        }
        
        if ([[[button titleLabel] text] isEqualToString:@"Fire"])
        {
            [button setTag:OE2600ButtonFire1];
        } else if ([[[button titleLabel] text] isEqualToString:@"Select"]) {
            [button setTag:OE2600ButtonSelect];
        } else if ([[[button titleLabel] text] isEqualToString:@"Reset"]) {
            [button setTag:OE2600ButtonReset];
        }
    }
    
    self.startButton.tag  = OE2600ButtonReset;
    self.selectButton.tag = OE2600ButtonSelect;
}

- (void)dPad:(JSDPad *)dPad didPressDirection:(JSDPadDirection)direction
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    
    [stellaCore didRelease2600Button:OE2600ButtonUp forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonDown forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonLeft forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonRight forPlayer:0];
    
    switch (direction)
    {
        case JSDPadDirectionUpLeft:
            [stellaCore didPush2600Button:OE2600ButtonUp forPlayer:0];
            [stellaCore didPush2600Button:OE2600ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionUp:
            [stellaCore didPush2600Button:OE2600ButtonUp forPlayer:0];
            break;
        case JSDPadDirectionUpRight:
            [stellaCore didPush2600Button:OE2600ButtonUp forPlayer:0];
            [stellaCore didPush2600Button:OE2600ButtonRight forPlayer:0];
            break;
        case JSDPadDirectionLeft:
            [stellaCore didPush2600Button:OE2600ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionRight:
            [stellaCore didPush2600Button:OE2600ButtonRight forPlayer:0];
            break;
        case JSDPadDirectionDownLeft:
            [stellaCore didPush2600Button:OE2600ButtonDown forPlayer:0];
            [stellaCore didPush2600Button:OE2600ButtonLeft forPlayer:0];
            break;
        case JSDPadDirectionDown:
            [stellaCore didPush2600Button:OE2600ButtonDown forPlayer:0];
            break;
        case JSDPadDirectionDownRight:
            [stellaCore didPush2600Button:OE2600ButtonDown forPlayer:0];
            [stellaCore didPush2600Button:OE2600ButtonRight forPlayer:0];
            break;
        default:
            break;
    }
    
    [self vibrate];
}

- (void)dPadDidReleaseDirection:(JSDPad *)dPad
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    
    [stellaCore didRelease2600Button:OE2600ButtonUp forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonDown forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonLeft forPlayer:0];
    [stellaCore didRelease2600Button:OE2600ButtonRight forPlayer:0];
}

- (void)buttonPressed:(JSButton *)button
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    NSInteger tag = button.tag;
    [stellaCore didPush2600Button:tag forPlayer:0];
    [self vibrate];
}

- (void)buttonReleased:(JSButton *)button
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    NSInteger tag = button.tag;
    [stellaCore didRelease2600Button:tag forPlayer:0];
}

- (void)pressStartForPlayer:(NSUInteger)player
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    [stellaCore didPush2600Button:OE2600ButtonReset forPlayer:player];
}

- (void)releaseStartForPlayer:(NSUInteger)player
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    [stellaCore didRelease2600Button:OE2600ButtonReset forPlayer:player];
}

- (void)pressSelectForPlayer:(NSUInteger)player
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    [stellaCore didPush2600Button:OE2600ButtonSelect forPlayer:player];
}

- (void)releaseSelectForPlayer:(NSUInteger)player
{
    PVStellaGameCore *stellaCore = (PVStellaGameCore *)self.emulatorCore;
    [stellaCore didRelease2600Button:OE2600ButtonSelect forPlayer:player];
}

@end

File: Provenance/Controller/PVStellaControllerViewController.swift
//
//  PVStellaControllerViewController.swift
//  Provenance
//
//  Created by Joe Mattiello on 17/03/2018.
//  Copyright (c) 2018 Joe Mattiello. All rights reserved.
//

import PVStella

fileprivate extension JSButton {
    var buttonTag : PV2600Button {
        get {
            return PV2600Button(rawValue: tag)!
        }
        set {
            tag = newValue.rawValue
        }
    }
}

class PVStellaControllerViewController: PVControllerViewController<PVStellaGameCore> {
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        buttonGroup?.subviews.forEach {
            guard let button = $0 as? JSButton, let title = button.titleLabel?.text else {
                return
            }
            if title == "Fire" {
                button.buttonTag = .fire1
            }
            else if title == "Select" {
                button.buttonTag = .select
            }
            else if title == "Reset" {
                button.buttonTag = .reset
            }
        }
        
        startButton?.buttonTag = .reset
        selectButton?.buttonTag = .select
    }

    override func dPad(_ dPad: JSDPad, didPress direction: JSDPadDirection) {
        let stellaCore = emulatorCore
        stellaCore?.didRelease(.up, forPlayer: 0)
        stellaCore?.didRelease(.down, forPlayer: 0)
        stellaCore?.didRelease(.left, forPlayer: 0)
        stellaCore?.didRelease(.right, forPlayer: 0)
        switch direction {
            case .upLeft:
                stellaCore?.didPush(.up, forPlayer: 0)
                stellaCore?.didPush(.left, forPlayer: 0)
            case .up:
                stellaCore?.didPush(.up, forPlayer: 0)
            case .upRight:
                stellaCore?.didPush(.up, forPlayer: 0)
                stellaCore?.didPush(.right, forPlayer: 0)
            case .left:
                stellaCore?.didPush(.left, forPlayer: 0)
            case .right:
                stellaCore?.didPush(.right, forPlayer: 0)
            case .downLeft:
                stellaCore?.didPush(.down, forPlayer: 0)
                stellaCore?.didPush(.left, forPlayer: 0)
            case .down:
                stellaCore?.didPush(.down, forPlayer: 0)
            case .downRight:
                stellaCore?.didPush(.down, forPlayer: 0)
                stellaCore?.didPush(.right, forPlayer: 0)
            default:
                break
        }
        vibrate()
    }

    override func dPadDidReleaseDirection(_ dPad: JSDPad) {
        let stellaCore = emulatorCore
        stellaCore?.didRelease(.up, forPlayer: 0)
        stellaCore?.didRelease(.down, forPlayer: 0)
        stellaCore?.didRelease(.left, forPlayer: 0)
        stellaCore?.didRelease(.right, forPlayer: 0)
    }

    override func buttonPressed(_ button: JSButton) {
        let stellaCore = emulatorCore
        stellaCore?.didPush(button.buttonTag, forPlayer: 0)
        vibrate()
    }

    override func buttonReleased(_ button: JSButton) {
        let stellaCore = emulatorCore
        stellaCore?.didRelease(button.buttonTag, forPlayer: 0)
    }

    override func pressStart(forPlayer player: Int) {
        let stellaCore = emulatorCore
        stellaCore?.didPush(.reset, forPlayer: UInt(player))
    }

    override func releaseStart(forPlayer player: Int) {
        let stellaCore = emulatorCore
        stellaCore?.didRelease(.reset, forPlayer: UInt(player))
    }

    override func pressSelect(forPlayer player: Int) {
        let stellaCore = emulatorCore
        stellaCore?.didPush(.select, forPlayer: UInt(player))
    }

    override func releaseSelect(forPlayer player: Int) {
        let stellaCore = emulatorCore
        stellaCore?.didRelease(.select, forPlayer: UInt(player))
    }
}
